Autogenerate blog post content and champion reports

nikomatsakis opened the pull request:
This extends the mdbook to include a "up to the minute" draft of the progress blog post. It also makes that look prettier. It's not perfect, some bits of the author comments don't yet manage links correct, but it's pretty good. It also removes the client-side rendering of progress bars, which wasn't really needed; I think I imagined doing more stuff on the client side but I don't really see the point now.

<img width="1690" height="3230" alt="image" src="https://github.com/user-attachments/assets/f5221608-5ccd-4640-9529-430a96d8ea01" />

nikomatsakis commented:
> 3000 line diff? That will take a while, I'll try a few commits at a time ^^.
> 
> There's a conflict to resolve though.

Eep, sorry =) it's a major overhaul to the structure. I think it'd be "ok" to just check it out and try it though

nikomatsakis commented:
Oh, also, I intentionally committing markdown files describing the plan so that you could read and vet the plan separately.

lqd commented:
No worries!

The first 6-7 commits have already landed in https://github.com/rust-lang/rust-project-goals/pull/417 and I think that's causing the git conflict.

lqd commented on templates/updates.hbs:
```diff
@@ -2,45 +2,21 @@ The Rust project is currently working towards a [slate of {{goal_count}} project
 
 ## Flagship goals
```
Do we still want to label the goals as flagship themselves here?

lqd commented on src/2025h2/relink-dont-rebuild.md:
```diff
@@ -1,10 +1,11 @@
+| Status           | Proposed                           |
+| Tracking issue   | [rust-lang/rust-project-goals#400] |
+| Zulip channel    |                                    |
+
 
```
The code isn't picking up champions when the sections are separated by empty lines like here. And that in turn makes the champion reports miss these goals of course.

(Also, shouldn't this goal be in the "faster builds" flagship themes?)

lqd commented on crates/rust-project-goals-cli/src/updates.rs:
```diff
@@ -37,7 +36,101 @@ pub(crate) fn generate_updates(
+                let task_owners_str = if doc.task_owners.is_empty() {
+                    "(none)".to_string()
+                } else {
+                    doc.task_owners.iter().cloned().collect::<Vec<_>>().join(", ")
```
The typesystem docs goal, and the evolving hierarchy goal, show an issue for task owners, here or in the goal document parsing (maybe due to the lack of "," separating some owners) or regexes -- but we can have duplicated task owners here. I'm not sure where we should be between structured parsing and flexibility in the text.

The RFL goal for compiler features data also looks incorrect here.

lqd reviewed (commented):
The new report structure is interesting. I'm not sure that the styling will work everywhere (project goals mdbook site, rust-lang blog, browser+mobile+light/dark themes). 

The markdown file helped a bit, though it looks out of date w/r/t what's implemented. 

<sub>the code itself is very, very meh</sub>

nikomatsakis commented:
I've cleaned up the git history (mostly) and addressed lqd's key feedback. There are still some parts of the code that are not as nice as I would like (fully qualified paths, kind of random or unhelpful comments) but I'm not going to address those at the moment because I really want the functionality here and they don't seem critical. Will come back to that.
